From: Thomas Klute <thomas2.klute@uni-dortmund.de>
Date: Tue, 13 Jan 2015 17:04:38 +0100
Subject: Check if filters exist before removing them in ssl_engine_disable

Trying to remove filters that are NULL leads to a segfault in the worker
thread. Check if c->input_filters and c->output_filters are defined
before removing and remove only if set.

Also, output filters should be removed with the dedicated function.
---
 src/mod_gnutls.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/mod_gnutls.c b/src/mod_gnutls.c
index e576fb8..0a32ffd 100644
--- a/src/mod_gnutls.c
+++ b/src/mod_gnutls.c
@@ -80,8 +80,10 @@ int ssl_engine_disable(conn_rec *c) {
     if(sc->enabled == GNUTLS_ENABLED_FALSE) {
         return 1;
     }
-    ap_remove_input_filter(c->input_filters);
-    ap_remove_input_filter(c->output_filters);
+    if (c->input_filters)
+        ap_remove_input_filter(c->input_filters);
+    if (c->output_filters)
+        ap_remove_output_filter(c->output_filters);
     mgs_cleanup_pre_config(c->pool);
     sc->enabled = 0;
     return 1;
